fdbclient/StorageServerInterface.h

```diff
@@ -741,6 +747,306 @@ struct ChangeFeedPopRequest {
+		Pending = 1,
+		Complete = 2,
+		Deleting = 3,
+		Fail = 4,
+	};
+
+	constexpr static FileIdentifier file_identifier = 13804342;
+	Version version;
+	KeyRange range;
+	int16_t format;
+	int16_t state;
+	UID checkpointID; // A unique id for this checkpoint.
+	UID ssID; // Storage server ID on which this checkpoint is created.
+	int referenceCount; // A reference count on the checkpoint, it can only be deleted when this is 0.
+
+	int64_t gcTime; // Time to delete this checkpoint, a Unix timestamp in seconds.
+
+	Optional<RocksDBColumnFamilyCheckpoint> rocksCF; // Present when format == RocksDBColumnFamily.
+
+	CheckpointMetaData() : format(InvalidFormat), state(InvalidState), referenceCount(0) {}
+	CheckpointMetaData(KeyRange const& range, CheckpointFormat format, UID const& ssID, UID const& checkpointID)
+	  : version(invalidVersion), range(range), format(format), ssID(ssID), checkpointID(checkpointID), state(Pending),
+	    referenceCount(0) {}
+	CheckpointMetaData(Version version, KeyRange const& range, CheckpointFormat format, UID checkpointID)
+	  : version(version), range(range), format(format), checkpointID(checkpointID), referenceCount(0) {}
+
+	CheckpointState getState() const { return static_cast<CheckpointState>(state); }
+
+	void setState(CheckpointState state) { this->state = static_cast<int16_t>(state); }
+
+	std::string toString() const {
+		std::string res = "Checkpoint MetaData:\nServer: " + ssID.toString() + "\nID: " + checkpointID.toString() +
+		                  "\nVersion: " + std::to_string(version) +
+		                  "\nFormat: " + getFdbCheckpointFormatName(static_cast<CheckpointFormat>(format)) +
+		                  "\nState: " + std::to_string(static_cast<int>(state)) + "\n";
+		if (rocksCF.present()) {
+			res += rocksCF.get().toString();
+		}
+		return res;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, version, range, format, state, checkpointID, ssID, gcTime, rocksCF);
+	}
+};
+
+// Request to search for a checkpoint for a minimum keyrange: `range`, at the specific version,
+// in the specific format.
+// A CheckpointMetaData will be returned if the specific checkpoint is found.
+struct GetCheckpointRequest {
+	constexpr static FileIdentifier file_identifier = 13804343;
+	Version version;
+	KeyRange range;
+	int16_t format;
+	Optional<UID> checkpointID; // When present, look for the checkpoint with the exact UID.
+	ReplyPromise<CheckpointMetaData> reply;
+
+	GetCheckpointRequest() {}
+	GetCheckpointRequest(Version version, KeyRange const& range, CheckpointFormat format)
+	  : version(version), range(range), format(format) {}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, version, range, format, checkpointID, reply);
+	}
+};
+
+// Reply to GetFileRequest, used to transfer files.
+struct GetFileReply : public ReplyPromiseStreamReply {
+	constexpr static FileIdentifier file_identifier = 13804345;
+	std::string name; // File name.
+	int64_t offset;
```

Comment:
Why is there a separate offset and size from StringRef's internal offset and size? If you only care about part of the file being transferred, wouldn't it be more efficient to only transfer that part of the file?

Reply:
It was intended for concurrent transfers, but this has been refactored to a more generic struct, these fields no longer exist.